internal/registry/azure/auth.go

```diff
@@ -0,0 +1,110 @@
+package azure
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/fluxcd/image-reflector-controller/internal/registry"
+)
+
+// Client is an Azure ACR client which can log into the registry and return
+// authorization information.
+type Client struct {
+	credential azcore.TokenCredential
+	scheme     string
+}
+
+// NewClient creates a new ACR client with default configurations.
+func NewClient() *Client {
+	return &Client{scheme: "https"}
+}
+
+// WithTokenCredential sets the token credential used by the ACR client.
+func (c *Client) WithTokenCredential(tc azcore.TokenCredential) *Client {
+	c.credential = tc
+	return c
+}
+
+// WithScheme sets the scheme of the http request that the client makes.
+func (c *Client) WithScheme(scheme string) *Client {
+	c.scheme = scheme
+	return c
+}
+
+// getLoginAuth returns authentication for ACR. The details needed for authentication
+// are gotten from environment variable so there is not need to mount a host path.
+func (c *Client) getLoginAuth(ctx context.Context, ref name.Reference) (authn.AuthConfig, error) {
+	var authConfig authn.AuthConfig
+
+	// Use default credentials if no token credential is provided.
+	// NOTE: NewDefaultAzureCredential() performs a lot of environment lookup
+	// for creating default token credential. Load it only when it's needed.
+	if c.credential == nil {
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return authConfig, err
+		}
+		c.credential = cred
+	}
+
+	// Obtain access token using the token credential.
+	armToken, err := c.credential.GetToken(ctx, policy.TokenRequestOptions{
+		Scopes: []string{string(arm.AzurePublicCloud) + ".default"},
```

Comment:
Is this only supported in public cloud? We can always extend this later, but it may be worth adding a comment to this effect.

Reply:
Added a TODO for this.